app/routes.py

```diff
@@ -1 +1,274 @@
+from sqlalchemy import asc, desc
+from datetime import datetime
+# from slack_sdk import WebClient
+# from slack_sdk.errors import SlackApiError
+import os
+import requests
+from app.models.goal import Goal
+import json
+
+
+
+
+tasks_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+goals_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+def validate_model(cls, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message":f"task {model_id} invalid"}, 400))
+
+    model = cls.query.get(model_id)
+    
+    if not model:
+        abort(make_response({"message":f"{cls.__name__} {model_id} not found"}, 404))
+    
+    return model
+
+
+@tasks_bp.route("", methods=["POST"])
+def create_task():
+    request_body = request.get_json()
+    try:
+        new_task = Task.from_dict(request_body)
+
+    except:
+        abort(make_response({"details": "Invalid data"}, 400))
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    
+    return make_response(jsonify({"task": new_task.to_dict()}),201)
+
+@tasks_bp.route("", methods=["GET"])
+
+def read_all_tasks():
+    title_query = request.args.get("title")
+    sort_filter = request.args.get("sort")
+    
+    if title_query:
+        tasks = Task.query.filter_by(title=title_query)
+    else:
+        tasks = Task.query.all()
+
+    if sort_filter:
+        if sort_filter == "asc":
+            tasks = Task.query.order_by(Task.title.asc())
+        elif sort_filter == "desc":
+            tasks = Task.query.order_by(Task.title.desc())
+    else:
+        tasks = Task.query.all()
+    
+    tasks_response = []
+    for task in tasks:
+        tasks_response.append(task.to_dict())
+    
+    return jsonify(tasks_response)
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+
+def read_one_task(task_id):
+    task = validate_model(Task, task_id)
+    if task.goal_id:
+        return {"task": task.to_dict_with_goal()}
+    else:
+        return {"task": task.to_dict()}
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model(Task, task_id)
+
+    request_body = request.get_json()
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+
+    db.session.commit()
+
+    return {"task": task.to_dict()}
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = validate_model(Task, task_id)
+
+    
+    db.session.delete(task)
+    db.session.commit()
+
+    return make_response({"details":f'Task {task.task_id} "{task.title}" successfully deleted'})
+
+def slack_notification(task):
+    token = os.environ.get("slack_token")
+    slack_url = "https://slack.com/api/chat.postMessage"
+    headers = {"Authorization":token}
+    body = {
+        "channel": "task-notifications",
+        "text": f"Someone just completed the task {task.title}", 
+    }
+
+    requests.post(slack_url, headers=headers, json=body)
+    return
```

Comment:
Love how you made this into a helper function! Make sure you are putting them a place that is easy to find, I typically put mine a separate folder or at the top of the file.